mlton/xml/parse-sxml.fun

```diff
@@ -0,0 +1,439 @@
+        | NONE   => false
+
+   fun isIdentFirst b = Char.isAlpha b orelse b = #"'"
+   fun isIdentRest b = Char.isAlphaNum b orelse b = #"'" orelse b = #"_" orelse b = #"."
+
+   val skipComments = T.optional(
+      T.string "(*" *> T.cut (T.manyCharsFailing(T.string "*)") *> T.string "*)" 
+         <|> T.failCut "Closing comment")
+      ) *> T.next
+
+   val space = T.sat(T.next, Char.isSpace) 
+   val spaces = T.many(space)
+   fun token s = T.notFollowedBy
+      (T.string s,
+       (T.char #"_") <|> (T.sat(T.next,Char.isAlphaNum))) <* spaces
+   fun symbol s = T.notFollowedBy
+      (T.string s,
+       (T.sat(T.next,fn b => isInfixChar b orelse b = #"_"))) <* spaces
+
+   val clOptions = T.manyCharsFailing(T.string "Datatypes:")
+
+   fun 'a makeNameResolver(f: string -> 'a): string -> 'a =
+      let
+         val hash = String.hash 
+         val map = HashSet.new{hash= hash o #1}
+         fun eq x (a: string * 'a) = String.equals(x, #1 a)
+      in
+         fn x => (#2 o HashSet.lookupOrInsert)(map, hash x, eq x, fn () => (x, f x))
+      end
+
+
+   val ident = (T.failing (token "in" <|> token "val" <|> token "fn" <|> token
+   "_" <|> token "=>" <|> token "case" <|> token "prim") *>
+      String.implode <$> (T.any
+         [(op ::) <$$>
+             (T.sat(T.next, isIdentFirst),
+              T.many (T.sat(T.next, isIdentRest))),
+          List.append <$$>
+             (T.many1 (T.sat(T.next, isInfixChar)),
+              (op ::) <$$> (T.char #"_", T.many (T.sat(T.next, Char.isDigit))) <|> T.pure []
+              (* just for collecting _0 *)
+              )])) <|> T.failCut "identifier"
+
+       
+
+   (* parse a tuple of parsers which must begin with a paren but may be unary *)
+   fun tupleOf p = Vector.fromList <$>
+      (T.char #"(" *> T.sepBy(spaces *> p, T.char #",") <* T.char #")")
+
+   fun vectorOf p = Vector.fromList <$>
+      (T.char #"[" *> T.sepBy(spaces *> p, T.char #",") <* T.char #"]")
+
+   fun doneRecord' () = T.char #"{" <* T.many(T.delay doneRecord' <|> T.failing(T.char #"}") *> T.next) <* T.char #"}"
+   val doneRecord = doneRecord' ()
+   fun fromRecord name p = T.peek 
+      (T.char #"{" *> T.many (() <$ T.delay doneRecord' <|> () <$ T.failing (token name <* symbol "=") <* T.next)
+       *> token name *> symbol "=" *> p)
+
+   fun casesOf(con, left, right) = Vector.fromList <$> T.sepBy1
+      (left <* spaces <* token "=>" >>= (fn l =>  
+         right >>= (fn r => con (l, r))),
+       spaces)
+   
+   fun optionOf p = SOME <$> (token "Some" *> T.cut(p)) <|> NONE <$ token "None"
+
+
+   (* too many arguments for the maps, curried to use <*> instead *)
+   fun makeTyp resolveTycon (args, ident) =
+      let
+         fun makeNullary() =
+            case ident of
+               "bool" => Type.bool
+             | "cpointer" => Type.cpointer
+             | "intInf" => Type.intInf
+             | "thread" => Type.thread
+             | "unit" => Type.unit
+             | "word8" => Type.word8
```

Comment:
Might need a more general solution to handle the non-primitive word sizes (e.g., `word7`).